src/charm.py

```diff
@@ -32,140 +31,144 @@
+            a Pebble layer specification for the Prometheus workload container.
+        """
+        logger.debug("Building pebble layer")
+        layer_config = {
+            "summary": "Prometheus layer",
+            "description": "Pebble layer configuration for Prometheus",
+            "services": {
+                self._name: {
+                    "override": "replace",
+                    "summary": "prometheus daemon",
+                    "command": self._command(),
+                    "startup": "enabled",
+                }
+            },
+        }
+
+        return Layer(layer_config)
+
+    @property
+    def _external_url(self) -> str:
+        """Return the external hostname to be passed to ingress via the relation.
+
+        If we do not have an ingress, then use the pod ip as hostname.
+        The reason to prefer this over the pod name (which is the actual
+        hostname visible from the pod) or a K8s service, is that those
+        are routable virtually exclusively inside the cluster (as they rely)
+        on the cluster's DNS service, while the ip address is _sometimes_
+        routable from the outside, e.g., when deploying on MicroK8s on Linux.
+        """
+        if web_external_url := self.model.config.get("web_external_url"):
+            return web_external_url
+        if ingress_url := self.ingress.url:
+            return ingress_url
+        return f"http://{socket.getfqdn()}:{self._port}"
+
+    def _patch_kubernetes(self):
+        self.service_patch = KubernetesServicePatch(self, [(f"{self.app.name}", self._port)])
+        self.resources_patch = KubernetesComputeResourcesPatch(
+            self,
+            self._name,
+            resource_reqs_func=lambda: self._resource_reqs_from_config(),
         )
-        self.framework.observe(self.on.update_status, self._update_status)
-        self.framework.observe(self.resources_patch.on.patch_failed, self._on_k8s_patch_failed)
+
+    def _register_event_handlers(self):
+        for event, handler in [
+            (self.on.prometheus_pebble_ready, self._configure),
+            (self.on.config_changed, self._configure),
+            (self.on.upgrade_charm, self._configure),
+            (self.ingress.on.ready_for_unit, self._on_ingress_ready),
+            (self.ingress.on.revoked_for_unit, self._on_ingress_revoked),
+            (self.on.receive_remote_write_relation_created, self._configure),
+            (self.on.receive_remote_write_relation_changed, self._configure),
+            (self.on.receive_remote_write_relation_broken, self._configure),
+            (self.metrics_consumer.on.targets_changed, self._configure),
+            (self.alertmanager_consumer.on.cluster_changed, self._configure),
+            (self.on.validate_configuration_action, self._on_validate_configuration),
+            (self.on.update_status, self._update_status),
+            (self.resources_patch.on.patch_failed, self._on_k8s_patch_failed),
+        ]:
```

Comment:
At that point I'm not sure whether I think it adds any value at all. In a world where we (hopefully) get rid of the common entrypoint, it would turn each line (currently) into three lines, defeating the purpose of tidying up. Also, nested for loops. 😩

Reply:
Reverted.